### PR TITLE
expression: rewrite builtin function: EXPORT_SET

### PR DESCRIPTION
**Pull request by zz-jason**
to #4080

### CODE REVIEW

**Comment by zz-jason** [comment]
@XuHuaiyu @breeswish @jackysp PTAL

**Comment by jackysp** [comment]
LGTM

File: expression/builtin_string.go [hunk 1]
```diff
@@ -2372,62 +2372,62 @@ func (c *exportSetFunctionClass) getFunction(ctx context.Context, args []Express
...
+	argTps = append(argTps, tpInt, tpString, tpString)
+	if len(args) > 3 {
+		argTps = append(argTps, tpString)
+	}
+	if len(args) > 4 {
+		argTps = append(argTps, tpInt)
+	}
+	bf := newBaseBuiltinFuncWithTp(args, ctx, tpString, argTps...)
+	bf.tp.Flen = mysql.MaxBlobWidth
+	sig := &builtinExportSetSig{baseStringBuiltinFunc{bf}}
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: split it into 3 sigs.

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

File: expression/builtin_string.go [hunk 2]
```diff
@@ -2369,78 +2371,143 @@ type exportSetFunctionClass struct {
...
+	case 5:
+		sig = &builtinExportSet5ArgSig{baseStringBuiltinFunc{bf}}
 	}
-	var result string
-	for i := 0; i < numberOfBits; i++ {
-		if bits&1 > 0 {
+	return sig.setSelf(sig), nil
+}
+
+func exportSet(bits int64, on, off, separator string, numberOfBits int64) string {
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: add a comment for this func.

File: expression/builtin_string.go [hunk 3]
```diff
@@ -2369,78 +2371,143 @@ type exportSetFunctionClass struct {
...
+	}
+
+	return exportSet(bits, on, off, ",", 64), false, nil
+}
+
+type builtinExportSet4ArgSig struct {
+	baseStringBuiltinFunc
+}
+
+// evalString evals EXPORT_SET(bits,on,off[,separator[,number_of_bits]]).
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: update comments.
so as the others

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

**Pull request merged**